Description: Don't redirect to the server's canonical name

I've had to patch mod_ucam_webauth for use on the SRCF, and thought I should 
explain why this has been necessary in case you want to incorporate my 
patch.  Fair warning: I've essentially had to _remove_ a feature!

The SRCF hosts a large number of Apache virtual hosts, and these are 
configured with UseCanonicalName turned off 
(http://httpd.apache.org/docs/2.2/mod/core.html#usecanonicalname).  This is 
because the canonical ServerName of any given host (usually e.g. 
spqr2.user.srcf.net) is not necessarily what the owner of that site wants in 
their URLs (e.g. if they have their own prettier domain name).  Indeed 
several users use mod_rewrite to ensure that the site is only served from 
their own domain name, redirecting from any others at which clients may 
arrive -- including from the canonical ServerName.

However, mod_ucam_webauth 1.4.3 second-guesses UseCanonicalName, forcing 
redirection to the canonical name even if UseCanonicalName is off.  This can 
in some situations cause a redirect loop, e.g. if mod_ucam_webauth and a 
user's configuration of mod_rewrite disagree over which URL the user should 
use.

Furthermore I'm not sure the forced redirection to the canonical ServerName 
is necessary at all (any more?).  Comments in the mod_ucam_webauth source 
seem to justify it by noting that if a user switches to another hostname for 
the same virtual host, the Raven cookie will be lost.  I would argue that a 
user is unlikely to change hostnames once they have arrived at a site, and 
if they do it is not a problem to force them to interact again with Raven 
(if they are forced to interact at all -- this depends on their preferences 
in Raven for automatic authentication to new sites, and I think the defaults 
for Raven accounts are set such that some reauthentications will happen 
silently).

The SRCF has been running with the relevant parts of mod_ucam_webauth 
disabled for the best part of a year and we have seen no problems with this 
setup, however I'd be interested to hear your thoughts.


Author: Malcolm Scott <debianpkg@malc.org.uk>
Last-Update: 2012-05-18

---

--- libapache2-mod-ucam-webauth-1.4.3.orig/mod_ucam_webauth.c
+++ libapache2-mod-ucam-webauth-1.4.3/mod_ucam_webauth.c
@@ -1085,29 +1085,13 @@ get_url(request_rec *r)
      depending on the presence (or otherwise) of ServerName and/or
      Port and/or Listen directive. Needs testing. */ 
 
-  char *url, *result;
+  char *url;
   apr_uri_t uri;
 
   url = ap_construct_url(r->pool, r->unparsed_uri, r);
   APACHE_LOG1(APLOG_DEBUG, "get_url: raw url = %s", url);
 
-  /* ap_construct_url honours UseCannonicalName but we really don't
-     want that so we re-parse this result and override the hostname
-     component with what we know we are really called
-  */
-
-#ifdef APACHE1_3
-  if (apr_uri_parse(r->pool, url, &uri) != HTTP_OK)
-    APACHE_LOG0(APLOG_CRIT, "Failed to parse own URL");
-#else
-  if (apr_uri_parse(r->pool, url, &uri))
-    APACHE_LOG0(APLOG_CRIT, "Failed to parse own URL");
-#endif
-  uri.hostname = r->server->server_hostname;
-  result = apr_uri_unparse(r->pool, &uri, (unsigned)0);
-
-  APACHE_LOG1(APLOG_DEBUG, "get_url: fixed url = %s", result);
-  return result;
+  return url;
 
 }
 
@@ -2565,31 +2549,6 @@ webauth_authn(request_rec *r)
     (APLOG_INFO, "** mod_ucam_webauth (%s) authn handler started for %s", 
      VERSION, r->uri);
 
-  /* If the hostname the user used (as reported by the 'Host' header)
-     doesn't match the configured hostname for this server then we are
-     going to have all sorts of problems with cookies and redirects,
-     so fix it (with a redirect) now. */
-
-  host = apr_pstrdup(r->pool,apr_table_get(r->headers_in, "Host"));
-  if (host != NULL) {
-    colon = strchr(host,':');
-    if (colon != NULL)
-      *colon = '\0';
-    if (r->server->server_hostname && 
-	strcasecmp(r->server->server_hostname,host)) {
-      colon = strchr(host,':');
-      if (colon != NULL)
-	*colon = '\0'; 
-      APACHE_LOG2
-	(APLOG_DEBUG,"Browser supplied hostname (%s) does not match "
-	 "configured hostname (%s) - redirecting",
-	 host, r->server->server_hostname);
-      apr_table_set(r->headers_out, "Location", get_url(r));
-      return (r->method_number == M_GET) ? 
-	HTTP_MOVED_TEMPORARILY : HTTP_SEE_OTHER;
-    }
-  }
-  
   c = (mod_ucam_webauth_cfg *) 
     ap_get_module_config(r->per_dir_config, &ucam_webauth_module);
   c = apply_config_defaults(r,c);
